api/channel.go

```diff
@@ -360,43 +360,60 @@ func JoinChannel(c *Context, channelId string, role string) {
 		user := uresult.Data.(*model.User)
 
-		if !c.HasPermissionsToTeam(channel.TeamId, "joinChannel") {
-			return
-		}
-
-		if channel.DeleteAt > 0 {
-			c.Err = model.NewAppError("joinChannel", "The channel has been archived or deleted", "")
-			c.Err.StatusCode = http.StatusBadRequest
+		if !c.HasPermissionsToTeam(channel.TeamId, "join") {
 			return
 		}
 
 		if channel.Type == model.CHANNEL_OPEN {
-			cm := &model.ChannelMember{ChannelId: channel.Id, UserId: c.Session.UserId,
-				Roles: role, NotifyProps: model.GetDefaultChannelNotifyProps()}
-
-			if cmresult := <-Srv.Store.Channel().SaveMember(cm); cmresult.Err != nil {
-				c.Err = cmresult.Err
-				return
-			}
-
-			post := &model.Post{ChannelId: channel.Id, Message: fmt.Sprintf(
-				`User %v has joined this channel.`,
-				user.Username), Type: model.POST_JOIN_LEAVE}
-			if _, err := CreatePost(c, post, false); err != nil {
-				l4g.Error("Failed to post join message %v", err)
-				c.Err = model.NewAppError("joinChannel", "Failed to send join request", "")
+			if _, err := AddUserToChannel(user, channel); err != nil {
+				c.Err = err
 				return
 			}
-
-			UpdateChannelAccessCacheAndForget(c.Session.TeamId, c.Session.UserId, channel.Id)
+			PostUserAddRemoveMessageAndForget(c, channel.Id, fmt.Sprintf(`User %v has joined this channel.`, user.Username))
 		} else {
-			c.Err = model.NewAppError("joinChannel", "You do not have the appropriate permissions", "")
+			c.Err = model.NewAppError("join", "You do not have the appropriate permissions", "")
 			c.Err.StatusCode = http.StatusForbidden
 			return
 		}
 	}
 }
 
+func PostUserAddRemoveMessageAndForget(c *Context, channelId string, message string) {
+	go func() {
+		post := &model.Post{
+			ChannelId: channelId,
+			Message:   message,
+			Type:      model.POST_JOIN_LEAVE,
+		}
+		if _, err := CreatePost(c, post, false); err != nil {
+			l4g.Error("Failed to post join message %v", err)
+		}
+	}()
+}
+
+func AddUserToChannel(user *model.User, channel *model.Channel) (*model.ChannelMember, *model.AppError) {
+	if channel.DeleteAt > 0 {
+		return nil, model.NewAppError("AddUserToChannel", "The channel has been archived or deleted", "")
+	}
+
+	if channel.Type != model.CHANNEL_OPEN && channel.Type != model.CHANNEL_PRIVATE {
+		return nil, model.NewAppError("AddUserToChannel", "Can not add user to this channel type", "")
+	}
+
+	newMember := &model.ChannelMember{ChannelId: channel.Id, UserId: user.Id, NotifyProps: model.GetDefaultChannelNotifyProps()}
+	if cmresult := <-Srv.Store.Channel().SaveMember(newMember); cmresult.Err != nil {
+		l4g.Error("Failed to add member user_id=%v channel_id=%v err=%v", user.Id, channel.Id, cmresult.Err)
+		return nil, model.NewAppError("AddUserToChannel", "Failed to add user to channel", "")
+	}
+
+	UpdateChannelAccessCacheAndForget(channel.TeamId, user.Id, channel.Id)
+
+	message := model.NewMessage(channel.TeamId, channel.Id, user.Id, model.ACTION_USER_ADDED)
+	PublishAndForget(message)
```

Comment:
I think `UpdateChannelAccessCacheAndForget` needs to be `UpdateChannelAccessCache`. There's a race condition here where if the message gets sent by the websocket before we're done updating the channel access cache, then the user won't get the websocket event since the cache thinks they don't yet have access to it.

Also couldn't hurt to throw the webscoket related stuff into a go routine since it doesn't need to complete before returning from the API service